util/ranger/detacher.go

```diff
@@ -201,23 +255,91 @@ func detachCNFCondAndBuildRangeForIndex(sctx sessionctx.Context, conditions []ex
 		shouldReserve: lengths[eqOrInCount] != types.UnspecifiedLength,
 	}
 	if considerDNF {
+		pointRes, offset, err := extractIndexPointRangesForCNF(sctx, conditions, cols, lengths)
+		if err != nil {
+			return nil, err
+		}
+		if pointRes != nil {
+			if len(pointRes.Ranges) == 0 {
+				return &DetachRangeResult{}, nil
+			}
+			if len(pointRes.Ranges[0].LowVal) > eqOrInCount {
+				eqOrInCount = len(pointRes.Ranges[0].LowVal)
+				newCols := cols[eqOrInCount:]
+				newLengths := lengths[eqOrInCount:]
+				tailConditions := append(conditions[:offset], conditions[offset+1:]...)
+				if len(newCols) == 0 || len(tailConditions) == 0 {
+					*res = *pointRes
+					res.RemainedConds = tailConditions
+					return res, nil
+				}
+				tailRes, err := DetachCondAndBuildRangeForIndex(sctx, tailConditions, newCols, newLengths)
+				if err != nil {
+					return nil, err
+				}
+				if len(tailRes.Ranges) == 0 {
+					return &DetachRangeResult{}, nil
+				}
+				res.Ranges = appendRanges2PointRanges(pointRes.Ranges, tailRes.Ranges)
+				res.AccessConds = append(pointRes.AccessConds, tailRes.AccessConds...)
+				res.RemainedConds = tailRes.RemainedConds
+				if pointRes.EqOrInCount == pointRes.EqCondCount {
+					res.EqCondCount = pointRes.EqCondCount + tailRes.EqCondCount
+				}
+				res.EqOrInCount = pointRes.EqOrInCount + tailRes.EqOrInCount
+				return res, nil
+			}
+		}
+		if eqOrInCount > 0 {
+			newCols := cols[eqOrInCount:]
+			newLengths := lengths[eqOrInCount:]
+			tailRes, err := DetachCondAndBuildRangeForIndex(sctx, newConditions, newCols, newLengths)
+			if err != nil {
+				return nil, err
+			}
+			if len(tailRes.Ranges) == 0 {
+				return &DetachRangeResult{}, nil
+			}
+			ranges, err = buildCNFIndexRange(stmtCtx, cols, tpSlice, lengths, eqOrInCount, accessConds)
+			if err != nil {
+				return nil, err
+			}
+			res.Ranges = appendRanges2PointRanges(ranges, tailRes.Ranges)
+			res.AccessConds = append(accessConds, tailRes.AccessConds...)
+			res.RemainedConds = append(filterConds, tailRes.RemainedConds...)
+			if res.EqOrInCount == res.EqCondCount {
+				res.EqCondCount = res.EqCondCount + tailRes.EqCondCount
+			}
+			res.EqOrInCount = res.EqOrInCount + tailRes.EqOrInCount
```

Comment:
How about creating a separate function for this logic?
It has the same logic as `lines276-289` for dealing with tail conditions.